### PR TITLE
[WIP][Qt] Add new column which uses the compatibility database to display …

### PR DESCRIPTION
**Pull request by scribam**
…the information per game

screenshot from 2017-09-10 19-08-15

### CODE REVIEW

**Comment by AniLeo** [comment]
It would probably be a good idea to have some kind of way for the user to know inside the emulator what the status means - people will see Loadable and think the game works for example.

File: rpcs3/rpcs3qt/game_list_frame.cpp [hunk 1]
```diff
@@ -925,7 +929,7 @@ void game_list_frame::resizeEvent(QResizeEvent *event)
 }
 
 /**
- Cleans and readds entries to table widget in UI.
+ Cleans and reads entries to table widget in UI.
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/game_list_frame.cpp]: :) this was no typo. But it's fine either way

**Comment by Megamouse** [comment]
maybe setText to something like: "Aquiring status" while it is in progress.
Or if no network available show: "Network unavailable"

**Comment by twdarkeh** [comment]
Maybe color the rows, or even just the cell with the status, according to the status? So make it green for playable, yellow for ingame, etc.

**Comment by rodjun** [comment]
Would be nice to show the RPCS3 version used on mouse over

**Comment by Megamouse** [comment]
Please reinitiate the compatibility on our Refresh button. I just started rpcs3 with wifi disabled and after re-enabling it there was no possibility of aquiring the values

**Comment by jjsat** [comment]
It would be good if the user could opt-out of this via option because not everyone will want the emulator to essentially transfer their game list.

**Comment by AniLeo** [comment]
"Transfer" how?

**Comment by jjsat** [comment]
Indirectly by accessing the compatibility database entries for all found games. Easy to extract the information even from the normal server log because I assume that the user agent from which this is accessed is identifiable. Even if not the access pattern should be pretty unique.

**Comment by Megamouse** [comment]
we can add a dialog that asks the user once if he wants it enabled or not,
And/or add it somewhere in the settings.

**Comment by AniLeo** [comment]
User agent is "Mozilla/5.0". No one is going to make any telemetry using the access logs, if we ever have telemetry it will be an opt-out setting. But I guess you can add it as an setting, why not.

**Comment by Asinin3** [comment]
Would be nice if you can see the status for different IDs because some IDs don't get updated. Possibly even right-click to go to the compat thread or search for it with the game id.

**Comment by Megamouse** [comment]
you can already go to compat with rightclick so there is no need for that in my opinion

**Comment by paulsapps** [comment]
What about also allowing the user to set the compatibility? Obviously this is a little more tricky to do but could be useful?

File: rpcs3/rpcs3qt/game_list_frame.cpp (hunk 1 above)
Comment (danilaml) [reply rpcs3/rpcs3qt/game_list_frame.cpp]: What typo? Shouldn't it stay as  `readds` as in `re-adds`?

Comment (scribam) [reply rpcs3/rpcs3qt/game_list_frame.cpp]: Oops, sorry about that. I reverted this change.

**Comment by scribam** [comment]
Thanks for your feedback everyone.

@AniLeo I added tooltips which display the status description from https://rpcs3.net/compatibility, is it ok for you?
@Megamouse I will think about adding more information while it is in progress. I added a more descriptive message when something goes wrong with the network. Also, the refresh button should work now as you expect.
@twdarkeh I agree, some colors would be nice. I will see what I can do.
@jjsat Compatibility is now disabled by default.

Where would you like to put the checkbox to handle the compatibility activation ? Should we put it in the welcome dialog as well with a text explaining the functionality and some details about privacy  ?

**Comment by rodjun** [comment]
Probably a better idea to just hide the compatibility tab when it's disabled. No point in showing something that just says "Disabled" on every line

**Comment by raveskirza** [comment]
May be the compatibility column should be visible all time otherwise the users will not know it exists.
We can have the compatibility option under menu, the same way we have the option to show Game list toolbar

File: rpcs3/rpcs3qt/GameCompatibility.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+private:
+	int compatibilityColumn;
+	int serialColumn;
+	QNetworkAccessManager *networkAccessManager;
+	QTableWidget *tableWidget;
+	const std::map<QString, QString> tooltips = {
+		{ "Playable", QObject::tr("Games that can be properly played from start to finish") },
+		{ "Ingame", QObject::tr("Games that go somewhere but not far enough to be considered playable") },
+		{ "Intro", QObject::tr("Games that only display some screens") },
+		{ "Loadable", QObject::tr("Games that only display some screens") },
```
Comment (MarioSonic2987) [inline rpcs3/rpcs3qt/GameCompatibility.h]: Intro and Loadable have the same description. It should be **_Games that display a black screen with an active framerate_** for Loadable status.

Comment (MarioSonic2987) [inline rpcs3/rpcs3qt/GameCompatibility.h]: Intro and Loadable have the same description. It should be **_Games that display a black screen with an active framerate_** for Loadable status.

**Comment by Renusek** [comment]
IMO it should be a separate column, but maybe colored game title (maybe you could disable that in options if you want all game titles to be black.

**Comment by Megamouse** [comment]
Just add some smileys as icons. no need for ugly colors

**Comment by twdarkeh** [comment]
As for where to put the checkbox to enable it, settings -> emulator -> UI settings, below "show welcome screen" would be the best place.

**Comment by digitaldude555** [comment]
Just have playable and not playable. Nothing/Loadable/Intro/Ingame is not relevant for users.

**Comment by Asinin3** [comment]
@digitaldude555 I don't see why not, they can open rpcs3 and see if there games have been making any progress. Also very useful for testers especially if it can show other IDs for the same game and the date it was tested on.

**Comment by rodjun** [comment]
@digitaldude555 Why have anything at all if people can just launch the game and find out that status

**Comment by scribam** [comment]
Colors and date added:

screenshot from 2017-09-11 23-30-55

File: rpcs3/rpcs3qt/GameCompatibility.h (hunk 2 above)
Comment (scribam) [reply rpcs3/rpcs3qt/GameCompatibility.h]: Fixed, thanks!

**Comment by greentop** [comment]
-Ubuntu 16.04 
(Compatibility manually enabled for testing.)

rpcs3 ubuntu 16 04 pr3426

**Comment by Luffykun007** [comment]
@scribam This build says that Qt5Network.dll is missing from my computer and does not work for me.

**Comment by danilaml** [comment]
>This build says that Qt5Network.dll is missing from my computer and does not work for me.

A proper illustration of why I think that https://github.com/RPCS3/rpcs3/pull/3432 is a better solution to the existing one.

**Comment by Asinin3** [comment]
@scribam Would it be possible to sort by last updated? Or maybe even status?

**Comment by raveskirza** [comment]
https://ci.appveyor.com/project/rpcs3/rpcs3/build/0.0.3-6154/artifacts
I get this error when I use the size scroll bar. Looks like the game list is refreshed for each step moved in the scroll bar. If you disable the compatibility, RPCS3 doesn't crash

image

**Comment by Megamouse** [comment]
yeah you should only refresh compatibility on certain refreshs, like the refresh button or first time loading.

**Comment by scribam** [comment]
@Luffykun007 Can you try again to run it under Windows ? I did a rebase which include #3432 and I hope it will fix the issue about Qt5Network.dll.

Also, there is now a checkbox to enable/disable the compatibility feature in the Settings window, Emulator tab.

**Comment by AniLeo** [comment]
Maybe we could have some kind of datetime comparing mechanism, if the game was tested after the build's datetime then warn the user that it may not work as they have an older build? 

On that note, having the build's compilation date on program title could be useful too so we don't have to check when is it from by looking for commit information.

**Comment by Asinin3** [comment]
@scribam I had the same issue as Luffy before, and I still have it with your new build.

**Comment by Megamouse** [comment]
@Asinin3 you of course need to copy over the new files

**Comment by rodjun** [comment]
I still don't really like the idea of the colored circles, feels unnatural next to other UI elements. An option to disable just the circles would be appreciated even if it's just a hidden ini config without fancy UI checkboxes

**Comment by Luffykun007** [comment]
@Scribam works for me now but I am having this error creating SSL context when it will check the compatibility and the emulator crashes.

capturar

**Comment by Megamouse** [comment]
@Luffykun007 
This seems to be a Qt bug. But some say you have to copy some ssl dlls to the rpcs3 folder

File: rpcs3/rpcs3qt/game_compatibility.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+};
+
+struct game_compatibility_status
+{
+	QString m_text;
+	QString m_tooltip;
+	QString m_color;
+};
+
+const std::map<QString, game_compatibility_status> GAME_COMPATIBILITY_STATUS_DATA = {
```
Comment (danilaml) [inline rpcs3/rpcs3qt/game_compatibility.h]: Why make this global? It's only ever used in `cpp`. Could be moved there as well (maybe even in a anonymous namespace).

File: rpcs3/rpcs3qt/game_compatibility.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+{
+private:
+	std::unique_ptr<QPainter> m_painter;
+
+public:
+	game_compatibility_pixmap(QColor color) : QPixmap(16, 16)
+	{
+		fill(Qt::transparent);
+
+		m_painter = std::make_unique<QPainter>(this);
```
Comment (danilaml) [inline rpcs3/rpcs3qt/game_compatibility.h]: Does it really need a painter as a member? It's used only once here. Could be just a local var.

**Comment by danilaml** [comment]
Works for me with the exe from the latest appveyor build. No copying was required.

**Comment by Megamouse** [comment]
@danilaml that might be because you have Qt installed and in path

**Comment by danilaml** [comment]
@Megamouse yeah, you are correct. What bug though? As I understand, if you use SSL it should be provided by your system/yourself, since it's may be illegal to redistribute cryptography libs such as OpenSSL in some parts of the world.

File: rpcs3/rpcs3qt/game_compatibility.h (hunk 3 above)
Comment (scribam) [reply rpcs3/rpcs3qt/game_compatibility.h]: Moved to the cpp file, thanks.

File: rpcs3/rpcs3qt/game_compatibility.h (hunk 4 above)
Comment (scribam) [reply rpcs3/rpcs3qt/game_compatibility.h]: Not really indeed. Use a local var now, thanks.

**Comment by Megamouse** [comment]
@danilaml yeah I mistook the issue for open. It is already closed though. sorry. 
But the common "fix" is indeed to provide the SSL stuff yourself.

@scribam what did you change?
I compiled some stuff for you in discord btw

**Comment by scribam** [comment]
@Megamouse I did a rebase on master and took care of few things that danilaml reviewed.

**Comment by Asinin3** [comment]
Sorry, I didn't copy the dll over my fault. Anyway Now I'm getting this after enabling the feature and hitting refresh. 
image

**Comment by MelchiorGaspar** [comment]
The new column is not even displaying for me or even on the drop-down menu list... :( :/

**Comment by AniLeo** [comment]
It's better to use the API than a local database so we can get newer results in real time. Newer results will encourage people to update their builds if a warning is included when the build tested is more recent.

Maybe the results could be somewhat cached? Could be useful for people that have a lot of games perhaps, but then again it would have to ping the API again to check for updates.

**Comment by MelchiorGaspar** [comment]
cached would be good....  just add a context menu entry or a menu entry some where to refresh the cache..
O_o

**Comment by scribam** [comment]
This PR needs more work than I expected and I do not think I am skilled enough to find a proper way to integrate with the current code. Maybe I will work on it again, later, and submit a more mature PR. Until this time, feel free to use my code as you want to bring this feature to RPCS3 :grin:

**Pull request closed without merging**